File: valid_parentheses/valid_parentheses.go
package main

import (
	"fmt"
)


func isValid(s string) bool {
	vMap := map[rune]rune{
		'{': '}',
        '(': ')',
        '[': ']',
	}
	tStack := [] rune {}

	for _, char := range s  {
		if _, ok := vMap[char]; ok{
			tStack = append(tStack, char)
		} else {
              if len(tStack) == 0 {
              	return false
              }
            opener := tStack[len(tStack)-1]
            tStack = tStack[:len(tStack)-1]

            if vMap[opener] != char {
            	return false
            }
		}
	}

	return true
}

func main() {
	tests := []string{"()", "()[]{}", "(]", "([)]", "{[]}"}
	for _, test := range tests {
		fmt.Printf("%s: %v\n", test, isValid(test))
	}
}

File: valid_parentheses/valid_parentheses.py
#!/usr/bin/env python3

def valid_parentheses(s):
    v_map = {'{':'}',
             '(':')',
             '[':']'}
    t_stack = []

    for x in s:
        if x in v_map: # if its an opening paren
            t_stack.append(x)
        else:
            if not t_stack:
                return False
            opener = t_stack.pop()
            if v_map[opener] != x:
                return False

    return True


if __name__ == '__main__':
    s1 = "()"
    s2 = "()[]{}"
    s3 = "(]"
    s4 = "([)]"
    s5 = "{[]}"

    print(valid_parentheses(s3))
